alf/environments/carla_sensors.py

```diff
@@ -616,10 +618,124 @@ def render(self, display):
+                of the rendered points
+            size (int): size of the rendered point in terms of pixels
+        """
+
+        assert len(color) == 3, ("the color code should contain values for "
+                                 "[R, G, B] channels respectively")
+
+        point_cam = self._world_to_camera_image(world_point, rgb_img.shape[1],
+                                                rgb_img.shape[0])
+
+        half_size = size // 2
+        for i in range(point_cam.shape[0]):
+            pt = point_cam[i]
+
+            xi = int(np.asscalar(pt[0]))
+            yi = int(np.asscalar(pt[1]))
+
+            if xi >= 0 and xi < rgb_img.shape[
+                    1] and yi >= 0 and yi < rgb_img.shape[0]:
+
+                xst = xi - half_size
+                yst = yi - half_size
+
+                xed = xst + size
+                yed = yst + size
+
+                xst = np.clip(xst, 0, rgb_img.shape[1])
+                xed = np.clip(xed, 0, rgb_img.shape[1])
+                yst = np.clip(yst, 0, rgb_img.shape[0])
+                yed = np.clip(yed, 0, rgb_img.shape[0])
+
+                rgb_img[yst:yed, xst:xed] = color
+
+    def _world_to_camera_image(self, world_points, image_width, image_height):
+        """Project points in world coordinates to camera image plane.
+
+        Adapted from https://github.com/carla-simulator/carla/blob/master/PythonAPI/examples/lidar_to_camera.py
+        Additional reference:
+        https://web.stanford.edu/class/cs231a/course_notes/01-camera-models.pdf
+
+        Args:
+            world_points (np.ndarray): [N, 3] in world coordinate
+            image_width (int): the width of the image to be rendered on
+            image_height (int): the height of the image to be rendered on
+        Output:
+            np.ndarray: representing the image plane coordinates for the set
+                of points that are visible in the camera. Its shape is [N', 3],
+                with N' <= N.
+        """
+
+        # Build the projection matrix K:
+        # K = [[Fx,  0, image_w/2],
+        #      [ 0, Fy, image_h/2],
+        #      [ 0,  0,         1]]
+
+        focal = image_width / (2.0 * np.tan(self._fov * np.pi / 360.0))
+
+        # In this case Fx and Fy are the same since the pixel aspect
+        # ratio is 1
+        K = np.identity(3)
+        K[0, 0] = K[1, 1] = focal
+        K[0, 2] = image_width / 2.0
+        K[1, 2] = image_height / 2.0
+
+        # [4, 4]
+        world_2_camera = np.linalg.inv(
+            _get_transform_matrix(self._sensor.get_transform()))
+
+        # transform the points from world space to camera space through the
+        # following several steps
+
+        # [N, 3] -> [3, N]
+        world_points = np.transpose(world_points, (1, 0))
+        # [3, N] -> [4, N]
+        world_points = np.concatenate(
+            (world_points, np.ones_like(world_points[0:1, ...])), axis=0)
+
+        sensor_points = np.dot(world_2_camera, world_points)
```

Comment:
is np.dot same as np.matmul here?

Reply:
Yes they are the same as the arguments are 2D arrays. 
Changed ```np.dot``` to ```np.matmul```, as the latter one is preferred according to np documents for 2D case.